### PR TITLE
Adding metahash storage (IPFS)

### PR DESCRIPTION
**Pull request by leonprou**
- [x] Adding medatahash to CC contract
- [x] Adding update method to CurrencyFactory
- [x] Adding tests for Currency and Issuance factories
- [x] Deploying contracts to ropsten

### CODE REVIEW

File: contracts/ColuLocalCurrency.sol [hunk 1]
```diff
@@ -12,21 +12,30 @@ contract ColuLocalCurrency is Ownable, Standard677Token, TokenHolder {
...
+
     /// @dev cotract to use when issuing a CC (Local Currency)
     /// @param _name string name for CC token that is created.
     /// @param _symbol string symbol for CC token that is created.
     /// @param _decimals uint8 percison for CC token that is created.
-    /// @param _totalSupply uint256 total supply of the CC token that is created. 
-    function ColuLocalCurrency(string _name, string _symbol, uint8 _decimals, uint256 _totalSupply) public {
-        require(_totalSupply != 0);     
+    /// @param _totalSupply uint256 total supply of the CC token that is created.
+    function ColuLocalCurrency(string _name, string _symbol, uint8 _decimals, uint256 _totalSupply, string _metadata) public {
```
Comment (tal-beja) [inline contracts/ColuLocalCurrency.sol]: add metadata field to the docs

File: contracts/CurrencyFactory.sol [hunk 2]
```diff
@@ -58,13 +58,15 @@ contract CurrencyFactory is Standard223Receiver, TokenHolder {
   /// @param _symbol string symbol for CC token that is created.
   /// @param _decimals uint8 percison for CC token that is created.
   /// @param _totalSupply uint256 total supply of the CC token that is created.
+  /// @param _metadata string IPFS hash for the CC token data.
```
Comment (tal-beja) [inline contracts/CurrencyFactory.sol]: The metadata field is a uri to the metadata, not necessary IPFS

File: contracts/CurrencyFactory.sol [hunk 3]
```diff
@@ -151,6 +153,17 @@ contract CurrencyFactory is Standard223Receiver, TokenHolder {
   	return (clnAddress == _token || currencyMap[_token].totalSupply > 0);
   }
 
+  /// @dev helper function to set metadata for the currency
+  /// @param _token address address of the token to update
+  /// @param _metadata string hash of the metadata of the token, this
+  /// hash can be accessed through IPFS
```
Comment (tal-beja) [inline contracts/CurrencyFactory.sol]: again, not necessary IPFS

File: contracts/IssuanceFactory.sol [hunk 4]
```diff
@@ -88,45 +88,47 @@ contract IssuanceFactory is CurrencyFactory {
 	/// @param _symbol string symbol of the token
 	/// @param _decimals uint8 ERC20 decimals of local currency
 	/// @param _totalSupply uint total supply of the local currency
+	/// @param _metadata string IPFS hash for the CC token data.
```
Comment (tal-beja) [inline contracts/IssuanceFactory.sol]: again

File: contracts/IssuanceFactory.sol [hunk 5]
```diff
@@ -88,45 +88,47 @@ contract IssuanceFactory is CurrencyFactory {
...
   /// @param _startTime uint256 blocktime for sale start
-  /// @param _endTime uint256 blocktime for sale end
+  /// @param _durationTime uint256 seconds for sale's durationTime
   /// @param _hardcap uint256 sale hardcap
   /// @param _reserveAmount uint256 sale softcap
   /// @param _targetPrice uint256 sale CC price per CLN if it were to pass the softcap
   function addToMap(address _token,
                     uint256 _startTime,
-                    uint256 _endTime,
+                    uint256 _durationTime,
```
Comment (tal-beja) [inline contracts/IssuanceFactory.sol]: why the change?

File: contracts/CurrencyFactory.sol (hunk 2 above)
Comment (leonprou) [reply contracts/CurrencyFactory.sol]: Now it's a hash, not an uri.

File: contracts/IssuanceFactory.sol (hunk 4 above)
Comment (leonprou) [reply contracts/IssuanceFactory.sol]: It can be everything, but I want to say that the convention is to save it to IPFS

File: contracts/CurrencyFactory.sol (hunk 2 above)
Comment (leonprou) [reply contracts/CurrencyFactory.sol]: We can make somethink like ERC721 suggestion:
https://github.com/ethereum/eips/issues/721 (tokenMedata of the second draft), or
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md

For now yes I just call it an URI

**Comment by leonprou** [comment]
I've found out that we not changing the owner of the CC when we open the the market (we just change the owner of the market maker. Talking with **Rotem** he explained that we want to keep the factory as the owner for further migrations. 

We've got to a conclusion to an issuer address to the CC contract. This field will be defined on a construction as an argument (and maybe a default of msg.sender).

Then the setMetadata of CC contract will have a issuerOnly modifer

This means we don't need (and can't use) setMetadata method in the factory.  Cause factory contract isn't the issuer.

**Comment by leonprou** [comment]
ERC721 doesn't provide a method to change the metadata, so it's immutable.

But developing this mutable metadata I was inspired by contracts like: 
https://github.com/Bounties-Network/StandardBounties/blob/master/contracts/StandardBounties.sol#L431

Also we can use this EIP: https://github.com/rareartlabs/EIPs/blob/master/EIPS/eip-erc20-metadata-extension.md. It extends ERC20 with a metadata field from ERC721

Another option is to make the tokenUri immutable, but the data in the Uri can be mutable. For example IPFS uri is immutable, while IPNS/HTTP uris are mutable.

**Comment by leonprou** [comment]
Opened a new branch, follow up here - https://github.com/colucom/CLN-solidity/pull/30
